felix/bpf-gpl/nat_lookup.h

```diff
@@ -58,9 +58,18 @@ static CALI_BPF_INLINE struct calico_nat_dest* calico_nat_lookup(ipv46_addr_t *i
 	}
 
 	if (!nat_lv1_val) {
-		struct cali_rt *rt;
+		struct cali_rt *rt = cali_rt_lookup(ip_dst);
 
 		CALI_DEBUG("NAT: Miss.\n");
+
+		/* If the traffic is from outside the cluster, to a blocked cidr, drop the packet to
+		 * prevent the packet from looping.
+		 */
+		if (CALI_F_FROM_HEP && !from_tun && rt && cali_rt_is_blackhole(rt)) {
```

Comment:
> If the traffic is from outside the cluster

This `CALI_F_FROM_HEP && !from_tun` does not mean that the traffic from outside the cluster. It only means it is not from this node and did not come through the tunnel. In fact, traffic coming through the tunnel is most likely from outside the cluster. To make sure that traffic is from outside the cluster, you need to check `ip_src` for route miss.

This branch in tc.c handle the case where we had NAT miss. I wonder if it would make more sense to do the blackhole lookup there or somewhere we already do a route lookup so that we do not need to do it 2x https://github.com/projectcalico/calico/blob/master/felix/bpf-gpl/tc.c#L424

Reply:
Perhaps here, because we already do the dst lookup for anyway. https://github.com/projectcalico/calico/blob/master/felix/bpf-gpl/tc.c#L545